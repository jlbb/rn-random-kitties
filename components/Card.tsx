import { StyleSheet, Text, View, Image } from "react-native";
import React, { Ref } from "react";

import ButtonRK, { PRIMARY, DEFAULT, ButtonRKTypes } from "./ButtonRK";
import { colors } from "../config/contants";

// import catImage from "../public/ey_human.jpg";

interface CardProps {
  children?: any;
  controls?: boolean;
  controlsItem?: {
    label: string;
    action: () => any;
  }[];
  imageSrc?: string;
  title: string;
}

const defaultCardProps: any = {
  children: "Lorem impsum sin dolor",
  controls: false,
  controlsItem: [
    {
      label: "Button1",
      action: () => null
    }
  ],
  title: "Card cat",
  imageSrc: ""
};

const Card = React.forwardRef(
  (
    {
      children: text = defaultCardProps.children,
      controls = defaultCardProps.controls,
      controlsItem = defaultCardProps.controlsItem,
      imageSrc = defaultCardProps.imageSrc,
      title = defaultCardProps.title
    }: CardProps,
    ref: Ref<any>
  ): JSX.Element => {
    const _renderOptionalControls = (): JSX.Element => {
      if (controls) {
        return (
          <View style={styles.controlsContainer}>
            {controlsItem.map((control, key) => {
              const buttonType: ButtonRKTypes = key === 0 ? DEFAULT : PRIMARY;

              return (
                <View
                  key={`key-cardcontrol-${key}`}
                  style={styles.controlButton}
                >
                  <ButtonRK
                    title={control.label}
                    type={buttonType}
                    onButtonPress={control.action}
                  />
                </View>
              );
            })}
          </View>
        );
      }
    };

    return (
      <View style={styles.container} ref={ref}>
        <Image style={styles.image} source={{ uri: imageSrc }} />
        <View style={styles.contentContainer}>
          <Text style={styles.title}>{title}</Text>
          <Text>{text}</Text>
          {_renderOptionalControls()}
        </View>
      </View>
    );
  }
);

const styles = StyleSheet.create({
  container: {
    flex: 1,
    flexBasis: "auto",
    alignItems: "center",
    backgroundColor: colors.WHITE_DEFAULT,
    justifyContent: "center",
    maxWidth: 300,
    width: "90%",
    marginTop: 15,
    marginBottom: 15
  },
  image: {
    flex: 1,
    width: "100%",
    minHeight: 150,
    backgroundColor: "slategray"
  },
  contentContainer: {
    flex: 1,
    padding: 15,
    width: "100%"
  },
  title: {
    fontSize: 24,
    fontWeight: "bold",
    marginBottom: 10
  },
  controlsContainer: {
    flexDirection: "row",
    justifyContent: "space-between",
    marginTop: 10
  },
  controlButton: {
    flex: 1,
    maxWidth: "100%",
    marginRight: 5,
    marginLeft: 5
  }
});

export default Card;
