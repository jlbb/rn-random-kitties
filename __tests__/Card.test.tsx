import Card from "../components/Card";
import { mount } from "enzyme";

describe("Card", () => {
  it("renders correctly", () => {
    const wrapper = mount(<Card title={"Test Card"} />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find("ButtonRK").length).toBe(0);
  });

  it("renders with 1 control by default", () => {
    const wrapper = mount(<Card title={"Test Card"} controls />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find("ButtonRK").length).toBeGreaterThanOrEqual(1);
  });

  it("renders the controls passed by props", () => {
    const wrapper = mount(
      <Card
        title={"Test Card"}
        controls
        controlsItem={[
          {
            label: "Buton1",
            action: jest.fn()
          },
          {
            label: "Buton2",
            action: jest.fn()
          }
        ]}
      />
    );

    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find("ButtonRK").length).toBe(2);
  });
});
