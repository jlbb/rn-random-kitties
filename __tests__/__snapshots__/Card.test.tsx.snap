// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card renders correctly 1`] = `
<ForwardRef
  title="Test Card"
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#ffffff",
        "flex": 1,
        "flexBasis": "auto",
        "justifyContent": "center",
        "marginBottom": 15,
        "marginTop": 15,
        "maxWidth": 300,
        "width": "90%",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#ffffff",
          "flex": 1,
          "flexBasis": "auto",
          "justifyContent": "center",
          "marginBottom": 15,
          "marginTop": 15,
          "maxWidth": 300,
          "width": "90%",
        }
      }
    >
      <Image
        source={
          Object {
            "uri": "",
          }
        }
        style={
          Object {
            "backgroundColor": "slategray",
            "flex": 1,
            "minHeight": 150,
            "width": "100%",
          }
        }
      >
        <Image
          source={
            Object {
              "uri": "",
            }
          }
          style={
            Object {
              "backgroundColor": "slategray",
              "flex": 1,
              "minHeight": 150,
              "width": "100%",
            }
          }
        />
      </Image>
      <View
        style={
          Object {
            "flex": 1,
            "padding": 15,
            "width": "100%",
          }
        }
      >
        <View
          style={
            Object {
              "flex": 1,
              "padding": 15,
              "width": "100%",
            }
          }
        >
          <Text
            style={
              Object {
                "fontSize": 24,
                "fontWeight": "bold",
                "marginBottom": 10,
              }
            }
          >
            <Text
              style={
                Object {
                  "fontSize": 24,
                  "fontWeight": "bold",
                  "marginBottom": 10,
                }
              }
            >
              Test Card
            </Text>
          </Text>
          <Text>
            <Text>
              Lorem impsum sin dolor
            </Text>
          </Text>
        </View>
      </View>
    </View>
  </View>
</ForwardRef>
`;

exports[`Card renders the controls passed by props 1`] = `
<ForwardRef
  controls={true}
  controlsItem={
    Array [
      Object {
        "action": [MockFunction],
        "label": "Buton1",
      },
      Object {
        "action": [MockFunction],
        "label": "Buton2",
      },
    ]
  }
  title="Test Card"
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#ffffff",
        "flex": 1,
        "flexBasis": "auto",
        "justifyContent": "center",
        "marginBottom": 15,
        "marginTop": 15,
        "maxWidth": 300,
        "width": "90%",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#ffffff",
          "flex": 1,
          "flexBasis": "auto",
          "justifyContent": "center",
          "marginBottom": 15,
          "marginTop": 15,
          "maxWidth": 300,
          "width": "90%",
        }
      }
    >
      <Image
        source={
          Object {
            "uri": "",
          }
        }
        style={
          Object {
            "backgroundColor": "slategray",
            "flex": 1,
            "minHeight": 150,
            "width": "100%",
          }
        }
      >
        <Image
          source={
            Object {
              "uri": "",
            }
          }
          style={
            Object {
              "backgroundColor": "slategray",
              "flex": 1,
              "minHeight": 150,
              "width": "100%",
            }
          }
        />
      </Image>
      <View
        style={
          Object {
            "flex": 1,
            "padding": 15,
            "width": "100%",
          }
        }
      >
        <View
          style={
            Object {
              "flex": 1,
              "padding": 15,
              "width": "100%",
            }
          }
        >
          <Text
            style={
              Object {
                "fontSize": 24,
                "fontWeight": "bold",
                "marginBottom": 10,
              }
            }
          >
            <Text
              style={
                Object {
                  "fontSize": 24,
                  "fontWeight": "bold",
                  "marginBottom": 10,
                }
              }
            >
              Test Card
            </Text>
          </Text>
          <Text>
            <Text>
              Lorem impsum sin dolor
            </Text>
          </Text>
          <View
            style={
              Object {
                "flexDirection": "row",
                "justifyContent": "space-between",
                "marginTop": 10,
              }
            }
          >
            <View
              style={
                Object {
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                  "marginTop": 10,
                }
              }
            >
              <View
                key="key-cardcontrol-0"
                style={
                  Object {
                    "flex": 1,
                    "marginLeft": 5,
                    "marginRight": 5,
                    "maxWidth": "100%",
                  }
                }
              >
                <View
                  style={
                    Object {
                      "flex": 1,
                      "marginLeft": 5,
                      "marginRight": 5,
                      "maxWidth": "100%",
                    }
                  }
                >
                  <ButtonRK
                    onButtonPress={[MockFunction]}
                    title="Buton1"
                    type="DEFAULT"
                  >
                    <View
                      style={
                        Object {
                          "flex": 1,
                        }
                      }
                    >
                      <View
                        style={
                          Object {
                            "flex": 1,
                          }
                        }
                      >
                        <TouchableHighlight
                          activeOpacity={0.85}
                          delayPressOut={100}
                          onPress={[Function]}
                          style={
                            Array [
                              Object {
                                "alignItems": "center",
                                "backgroundColor": "#984fa0",
                                "flexDirection": "row",
                                "height": "100%",
                                "justifyContent": "center",
                                "minHeight": 50,
                              },
                              Object {
                                "backgroundColor": "#984fa0",
                              },
                            ]
                          }
                          underlayColor="black"
                        >
                          <View
                            accessible={true}
                            isTVSelectable={true}
                            onResponderGrant={[Function]}
                            onResponderMove={[Function]}
                            onResponderRelease={[Function]}
                            onResponderTerminate={[Function]}
                            onResponderTerminationRequest={[Function]}
                            onStartShouldSetResponder={[Function]}
                            style={
                              Array [
                                Object {
                                  "alignItems": "center",
                                  "backgroundColor": "#984fa0",
                                  "flexDirection": "row",
                                  "height": "100%",
                                  "justifyContent": "center",
                                  "minHeight": 50,
                                },
                                Object {
                                  "backgroundColor": "#984fa0",
                                },
                              ]
                            }
                          >
                            <View
                              accessible={true}
                              isTVSelectable={true}
                              onResponderGrant={[Function]}
                              onResponderMove={[Function]}
                              onResponderRelease={[Function]}
                              onResponderTerminate={[Function]}
                              onResponderTerminationRequest={[Function]}
                              onStartShouldSetResponder={[Function]}
                              style={
                                Array [
                                  Object {
                                    "alignItems": "center",
                                    "backgroundColor": "#984fa0",
                                    "flexDirection": "row",
                                    "height": "100%",
                                    "justifyContent": "center",
                                    "minHeight": 50,
                                  },
                                  Object {
                                    "backgroundColor": "#984fa0",
                                  },
                                ]
                              }
                            >
                              <Text
                                style={
                                  Array [
                                    Object {
                                      "color": "#f5f5f5",
                                      "fontSize": 18,
                                      "textTransform": "uppercase",
                                    },
                                    Object {
                                      "color": "#f5f5f5",
                                      "fontSize": 16,
                                    },
                                  ]
                                }
                              >
                                <Text
                                  style={
                                    Array [
                                      Object {
                                        "color": "#f5f5f5",
                                        "fontSize": 18,
                                        "textTransform": "uppercase",
                                      },
                                      Object {
                                        "color": "#f5f5f5",
                                        "fontSize": 16,
                                      },
                                    ]
                                  }
                                >
                                  Buton1
                                </Text>
                              </Text>
                            </View>
                          </View>
                        </TouchableHighlight>
                      </View>
                    </View>
                  </ButtonRK>
                </View>
              </View>
              <View
                key="key-cardcontrol-1"
                style={
                  Object {
                    "flex": 1,
                    "marginLeft": 5,
                    "marginRight": 5,
                    "maxWidth": "100%",
                  }
                }
              >
                <View
                  style={
                    Object {
                      "flex": 1,
                      "marginLeft": 5,
                      "marginRight": 5,
                      "maxWidth": "100%",
                    }
                  }
                >
                  <ButtonRK
                    onButtonPress={[MockFunction]}
                    title="Buton2"
                    type="PRIMARY"
                  >
                    <View
                      style={
                        Object {
                          "flex": 1,
                        }
                      }
                    >
                      <View
                        style={
                          Object {
                            "flex": 1,
                          }
                        }
                      >
                        <TouchableHighlight
                          activeOpacity={0.85}
                          delayPressOut={100}
                          onPress={[Function]}
                          style={
                            Array [
                              Object {
                                "alignItems": "center",
                                "backgroundColor": "#984fa0",
                                "flexDirection": "row",
                                "height": "100%",
                                "justifyContent": "center",
                                "minHeight": 50,
                              },
                              Object {
                                "backgroundColor": "#229a80",
                              },
                            ]
                          }
                          underlayColor="black"
                        >
                          <View
                            accessible={true}
                            isTVSelectable={true}
                            onResponderGrant={[Function]}
                            onResponderMove={[Function]}
                            onResponderRelease={[Function]}
                            onResponderTerminate={[Function]}
                            onResponderTerminationRequest={[Function]}
                            onStartShouldSetResponder={[Function]}
                            style={
                              Array [
                                Object {
                                  "alignItems": "center",
                                  "backgroundColor": "#984fa0",
                                  "flexDirection": "row",
                                  "height": "100%",
                                  "justifyContent": "center",
                                  "minHeight": 50,
                                },
                                Object {
                                  "backgroundColor": "#229a80",
                                },
                              ]
                            }
                          >
                            <View
                              accessible={true}
                              isTVSelectable={true}
                              onResponderGrant={[Function]}
                              onResponderMove={[Function]}
                              onResponderRelease={[Function]}
                              onResponderTerminate={[Function]}
                              onResponderTerminationRequest={[Function]}
                              onStartShouldSetResponder={[Function]}
                              style={
                                Array [
                                  Object {
                                    "alignItems": "center",
                                    "backgroundColor": "#984fa0",
                                    "flexDirection": "row",
                                    "height": "100%",
                                    "justifyContent": "center",
                                    "minHeight": 50,
                                  },
                                  Object {
                                    "backgroundColor": "#229a80",
                                  },
                                ]
                              }
                            >
                              <Text
                                style={
                                  Array [
                                    Object {
                                      "color": "#f5f5f5",
                                      "fontSize": 18,
                                      "textTransform": "uppercase",
                                    },
                                    Object {
                                      "color": "#f5f5f5",
                                      "fontSize": 16,
                                    },
                                  ]
                                }
                              >
                                <Text
                                  style={
                                    Array [
                                      Object {
                                        "color": "#f5f5f5",
                                        "fontSize": 18,
                                        "textTransform": "uppercase",
                                      },
                                      Object {
                                        "color": "#f5f5f5",
                                        "fontSize": 16,
                                      },
                                    ]
                                  }
                                >
                                  Buton2
                                </Text>
                              </Text>
                            </View>
                          </View>
                        </TouchableHighlight>
                      </View>
                    </View>
                  </ButtonRK>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</ForwardRef>
`;

exports[`Card renders with 1 control by default 1`] = `
<ForwardRef
  controls={true}
  title="Test Card"
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#ffffff",
        "flex": 1,
        "flexBasis": "auto",
        "justifyContent": "center",
        "marginBottom": 15,
        "marginTop": 15,
        "maxWidth": 300,
        "width": "90%",
      }
    }
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "#ffffff",
          "flex": 1,
          "flexBasis": "auto",
          "justifyContent": "center",
          "marginBottom": 15,
          "marginTop": 15,
          "maxWidth": 300,
          "width": "90%",
        }
      }
    >
      <Image
        source={
          Object {
            "uri": "",
          }
        }
        style={
          Object {
            "backgroundColor": "slategray",
            "flex": 1,
            "minHeight": 150,
            "width": "100%",
          }
        }
      >
        <Image
          source={
            Object {
              "uri": "",
            }
          }
          style={
            Object {
              "backgroundColor": "slategray",
              "flex": 1,
              "minHeight": 150,
              "width": "100%",
            }
          }
        />
      </Image>
      <View
        style={
          Object {
            "flex": 1,
            "padding": 15,
            "width": "100%",
          }
        }
      >
        <View
          style={
            Object {
              "flex": 1,
              "padding": 15,
              "width": "100%",
            }
          }
        >
          <Text
            style={
              Object {
                "fontSize": 24,
                "fontWeight": "bold",
                "marginBottom": 10,
              }
            }
          >
            <Text
              style={
                Object {
                  "fontSize": 24,
                  "fontWeight": "bold",
                  "marginBottom": 10,
                }
              }
            >
              Test Card
            </Text>
          </Text>
          <Text>
            <Text>
              Lorem impsum sin dolor
            </Text>
          </Text>
          <View
            style={
              Object {
                "flexDirection": "row",
                "justifyContent": "space-between",
                "marginTop": 10,
              }
            }
          >
            <View
              style={
                Object {
                  "flexDirection": "row",
                  "justifyContent": "space-between",
                  "marginTop": 10,
                }
              }
            >
              <View
                key="key-cardcontrol-0"
                style={
                  Object {
                    "flex": 1,
                    "marginLeft": 5,
                    "marginRight": 5,
                    "maxWidth": "100%",
                  }
                }
              >
                <View
                  style={
                    Object {
                      "flex": 1,
                      "marginLeft": 5,
                      "marginRight": 5,
                      "maxWidth": "100%",
                    }
                  }
                >
                  <ButtonRK
                    onButtonPress={[Function]}
                    title="Button1"
                    type="DEFAULT"
                  >
                    <View
                      style={
                        Object {
                          "flex": 1,
                        }
                      }
                    >
                      <View
                        style={
                          Object {
                            "flex": 1,
                          }
                        }
                      >
                        <TouchableHighlight
                          activeOpacity={0.85}
                          delayPressOut={100}
                          onPress={[Function]}
                          style={
                            Array [
                              Object {
                                "alignItems": "center",
                                "backgroundColor": "#984fa0",
                                "flexDirection": "row",
                                "height": "100%",
                                "justifyContent": "center",
                                "minHeight": 50,
                              },
                              Object {
                                "backgroundColor": "#984fa0",
                              },
                            ]
                          }
                          underlayColor="black"
                        >
                          <View
                            accessible={true}
                            isTVSelectable={true}
                            onResponderGrant={[Function]}
                            onResponderMove={[Function]}
                            onResponderRelease={[Function]}
                            onResponderTerminate={[Function]}
                            onResponderTerminationRequest={[Function]}
                            onStartShouldSetResponder={[Function]}
                            style={
                              Array [
                                Object {
                                  "alignItems": "center",
                                  "backgroundColor": "#984fa0",
                                  "flexDirection": "row",
                                  "height": "100%",
                                  "justifyContent": "center",
                                  "minHeight": 50,
                                },
                                Object {
                                  "backgroundColor": "#984fa0",
                                },
                              ]
                            }
                          >
                            <View
                              accessible={true}
                              isTVSelectable={true}
                              onResponderGrant={[Function]}
                              onResponderMove={[Function]}
                              onResponderRelease={[Function]}
                              onResponderTerminate={[Function]}
                              onResponderTerminationRequest={[Function]}
                              onStartShouldSetResponder={[Function]}
                              style={
                                Array [
                                  Object {
                                    "alignItems": "center",
                                    "backgroundColor": "#984fa0",
                                    "flexDirection": "row",
                                    "height": "100%",
                                    "justifyContent": "center",
                                    "minHeight": 50,
                                  },
                                  Object {
                                    "backgroundColor": "#984fa0",
                                  },
                                ]
                              }
                            >
                              <Text
                                style={
                                  Array [
                                    Object {
                                      "color": "#f5f5f5",
                                      "fontSize": 18,
                                      "textTransform": "uppercase",
                                    },
                                    Object {
                                      "color": "#f5f5f5",
                                      "fontSize": 16,
                                    },
                                  ]
                                }
                              >
                                <Text
                                  style={
                                    Array [
                                      Object {
                                        "color": "#f5f5f5",
                                        "fontSize": 18,
                                        "textTransform": "uppercase",
                                      },
                                      Object {
                                        "color": "#f5f5f5",
                                        "fontSize": 16,
                                      },
                                    ]
                                  }
                                >
                                  Button1
                                </Text>
                              </Text>
                            </View>
                          </View>
                        </TouchableHighlight>
                      </View>
                    </View>
                  </ButtonRK>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </View>
</ForwardRef>
`;
